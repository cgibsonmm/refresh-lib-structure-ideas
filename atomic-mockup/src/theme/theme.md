# Theme

This folder currently feels misused in my mind

My Idea here would be to move the components folder into the parent components folder since I feel like they are are atom components and not just extensions of MUI theme
