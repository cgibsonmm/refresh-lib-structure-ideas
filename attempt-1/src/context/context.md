# Context

This folder will house any context used in the refresh lib this will be mirrored in each brand app
