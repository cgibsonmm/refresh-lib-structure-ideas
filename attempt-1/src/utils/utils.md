# Utils

This folder will house global used Utils

I.E.

1. commaFormat
