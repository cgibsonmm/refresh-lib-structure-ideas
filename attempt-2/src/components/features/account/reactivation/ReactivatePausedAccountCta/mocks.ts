import { fetchAccountMock } from '@/account/account.mock';

export const pausedAccountMock = {
  ...fetchAccountMock,
  account: {
    ...fetchAccountMock.account,
    subscription_info: {
      ...fetchAccountMock.account.subscription_info,
      monthly_reports_remaining: 100,
      monthly_report_limit: 100,
    },
  },
  meta: {
    ...fetchAccountMock.meta,
    subscription_state: 'paused',
  },
};
