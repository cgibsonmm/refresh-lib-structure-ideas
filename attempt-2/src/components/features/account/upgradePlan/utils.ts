/**
 * Calculates the cost per month for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @returns Cost per month
 */
export function monthlyPlanCost(cost: number, renewalPeriod: number): number {
  return cost / renewalPeriod;
}

/**
 * Calculates the monthly cost per report for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @param limit Monthly report limit of the plan
 * @returns Cost per report rounded to the nearest hundredth
 */
export function costPerReport(
  cost: number,
  renewalPeriod: number,
  limit: number
): number {
  const mthlyPlanCost = Math.ceil(monthlyPlanCost(cost, renewalPeriod));

  // Do not change, the rounding used here is important and was agreed with the product team
  //round to hundredths
  const costPerReport = Math.ceil((mthlyPlanCost / limit) * 100) / 100;

  return costPerReport;
}

/**
 * Calculates the savings based on the monthly cost per report of upgrading from a base plan to a new plan
 *
 * @param baseCost Cost of the base plan
 * @param baseRenewalPeriod Number of months the base plan is valid for
 * @param baseLimit Monthly report limit of the base plan
 * @param upgradeCost Cost of the upgrade plan
 * @param upgradeRenewalPeriod Number of months the new plan is valid for
 * @param upgradeLimit Monthly report limit of the upgrade plan
 * @returns Savings as a percentage
 */
export function savings(
  baseCost: number,
  baseRenewalPeriod: number,
  baseLimit: number,
  upgradeCost: number,
  upgradeRenewalPeriod: number,
  upgradeLimit: number
): number {
  const baseCostPerReport = costPerReport(
    baseCost,
    baseRenewalPeriod,
    baseLimit
  );
  const upgradeCostPerReport = costPerReport(
    upgradeCost,
    upgradeRenewalPeriod,
    upgradeLimit
  );

  // Do not change, the rounding used here is important and was agreed with the product team
  const savings = Math.floor(
    ((baseCostPerReport - upgradeCostPerReport) / baseCostPerReport) * 100
  );

  return savings;
}
