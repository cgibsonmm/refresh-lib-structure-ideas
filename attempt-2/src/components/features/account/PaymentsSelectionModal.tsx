import { useContext, useState } from 'react';
import { Navigation, Pagination, Swiper } from 'swiper';
import { PaymentImages, PaymentMethods } from '@/account/Interfaces';
import { useSession } from '@/auth/index';
import { AppConstants, Constants } from '@/context/AppConstants';
import { Text, Button, Carousel, Modal, styled } from '@/theme/index';
import {
  useGetPaymentMethod,
  SetDefaultMethodPaymentTest,
} from './paymentMethodsService';
import {
  useProcessCard,
  useProcessPayPal,
} from './payments/useSubscriptionUpgrade';

const CarouselWrapper = styled('div')(({ theme }) => ({
  '.payments-swiper': {
    width: '80%',
    padding: `${theme.spacing(1.25)} ${theme.spacing(1.25)} ${theme.spacing(
      3.125
    )} ${theme.spacing(1.25)}`,
  },

  '.carousel-div-wrapper': {
    display: 'flex',
    justifyContent: 'center',
  },
  '.swiper-carousel': {
    width: '80%',
  },

  '.swiperSlide': {
    minHeight: `${theme.spacing(20)}`,
    padding: `${theme.spacing(1.25)}`,
    marginBottom: `${theme.spacing(2.5)}`,
    boxShadow: `${theme.spacing(0)} ${theme.spacing(0.5)} ${theme.spacing(
      1
    )} ${theme.spacing(0)} rgba(0, 0, 0, 0.2), ${theme.spacing(
      0
    )} ${theme.spacing(0.5)} ${theme.spacing(2.5)} ${theme.spacing(
      0
    )} rgba(0, 0, 0, 0.19);`,
    borderRadius: `${theme.spacing(1.25)}`,
  },
  '.default-text': {
    position: 'absolute',
    right: `${theme.spacing(1.25)}`,
    bottom: `${theme.spacing(1.25)}`,
  },
  '.text-support': {
    wordBreak: 'break-word',
    maxHeight: `${theme.spacing(8)}`,
    overflow: 'hidden',
  },
  [theme.breakpoints.down('sm')]: {
    '.payments-swiper': {
      padding: `${theme.spacing(1.25)} ${theme.spacing(2)} ${theme.spacing(
        3.125
      )} ${theme.spacing(2)}`,
    },
    '.swiperSlide': {
      height: 'auto',
      marginBottom: `${theme.spacing(2.25)}`,
      minHeight: `${theme.spacing(28.75)}`,
    },
    '.text-support': {
      fontSize: `${theme.spacing(1.5)}`,
      overflow: 'hidden',
    },
    '.default-text': {
      bottom: `${theme.spacing(1.25)}`,
    },
    '.carousel-div-wrapper': {
      display: 'block',
    },
    '.swiper-carousel': {
      width: '100%',
    },
  },
}));

const FlexDivContainer = styled('div')(({ theme }) => ({
  display: 'flex',
  gap: `${theme.spacing(1.25)}`,
  justifyContent: 'space-between',

  [theme.breakpoints.down('sm')]: {
    flexDirection: 'column',
    fontSize: `${theme.spacing(1.5)}`,
    alignItems: 'center',
  },
}));

const FlexDivCenter = styled('div')(({ theme }) => ({
  display: 'flex',
  justifyContent: 'center',
  textAlign: 'center',

  [theme.breakpoints.down('sm')]: {
    flexDirection: 'column',
    alignItems: 'center',
  },
}));

const CardColumn = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  alignContent: 'space-between',
  marginBottom: `${theme.spacing(1.25)}`,
  gap: `${theme.spacing(2.5)}`,
  maxWidth: `${theme.spacing(18.75)}`,

  img: {
    width: `${theme.spacing(12.5)}`,
    height: `${theme.spacing(7.5)}`,
  },

  '.text-height': {
    minHeight: `${theme.spacing(7.5)}`,
  },

  [theme.breakpoints.down('sm')]: {
    textAlign: 'center',
    alignItems: 'center',
    '.text-height': {
      minHeight: 'initial',
    },
  },
}));

const paymentImagesDictionary: { [key: string]: string } = {
  'Braintree::ApplePayCard': 'applePay',
  'Braintree::GooglePayCard': 'googlePay',
  'Braintree::VenmoAccount': 'venmo',
};

const displayBraintreeMethodName: { [key: string]: string } = {
  'Braintree::ApplePayCard': 'Apple Pay Account',
  'Braintree::GooglePayCard': 'Google Pay Account',
  'Braintree::VenmoAccount': 'Venmo Account',
};

const getSlideData = (
  slide: HTMLElement
): Pick<PaymentMethods, 'id' | 'payment_type'> => {
  const paymentTypes: { [key: string]: string } = {
    'Braintree::ApplePayCard': 'braintree',
    'Braintree::GooglePayCard': 'braintree',
    'Braintree::VenmoAccount': 'braintree',
    paypal: 'paypal',
    american_express: 'card',
    discover: 'card',
    master: 'card',
    visa: 'card',
  };

  if (!slide.dataset.id) throw new Error('Error selecting payment method id');

  const paymentType = slide.dataset.card_type || slide.dataset.braintree_type;

  if (!paymentType) throw new Error('Error selecting payment method type');

  const payment_type = paymentTypes[paymentType];

  return {
    id: parseInt(slide.dataset.id),
    payment_type: payment_type,
  };
};

const carouselElements = (
  { payment_methods }: { payment_methods: PaymentMethods[] },
  context: Constants
) => {
  return payment_methods.map((card: PaymentMethods, index: number) => {
    const isCreditCard = Boolean(card.expiration_month && card.expiration_year);
    const imgSrc =
      context[
        (card.card_type as keyof PaymentImages) ||
          (paymentImagesDictionary[card.braintree_type] as keyof PaymentImages)
      ];

    const paymentInformation = () => {
      if (card.card_type) {
        return isCreditCard ? card.description : card.email;
      }
      return displayBraintreeMethodName[card.braintree_type];
    };

    return (
      <div
        className="swiperSlide"
        data-id={card.id}
        data-card_type={card.card_type}
        data-braintree_type={card.braintree_type}
        key={index}
      >
        <FlexDivContainer>
          <CardColumn>
            <img src={imgSrc} alt="Card Logo" />
            {card.first_name && card.last_name && (
              <Text
                variant="caption"
                align="center"
                className="text-support"
              >{`${card.first_name} ${card.last_name}`}</Text>
            )}
          </CardColumn>
          <CardColumn>
            <Text variant="h5" margin={0} className="text-support text-height">
              {paymentInformation()}
            </Text>
            {isCreditCard && (
              <Text
                margin={0}
                variant="h5"
                className="text-support"
              >{`EXP ${card.expiration_month}/${card.expiration_year}`}</Text>
            )}
          </CardColumn>
        </FlexDivContainer>
        {card.active && (
          <Text
            variant="h2"
            align="right"
            className="text-support default-text"
          >
            DEFAULT
          </Text>
        )}
      </div>
    );
  });
};

/**
 * This component is a generic modal used to display a list of payment methods associated to the account.
 * It allows the user to select a payment method as default.
 */
export const PaymentsSelectionModal = ({
  open,
  cta = 'Make Default',
  inReactivation,
  isRecycling = false,
  onClose,
  onAddPaymentMethod,
  onSelectedPaymentMethod,
  defaultPaymentMethods,
}: {
  open: boolean;
  /** Custom text for the CTA button. */
  cta?: 'Make Default' | 'Use this Card' | (string & Record<never, never>);
  /**
   * Specifies if the modal is being used in the reactivation flow which changes the display
   * of the buttons from row to column.
   */
  inReactivation?: boolean;
  /**
   * Specifies if the modal is being used in the recycling flow which makes the main button inmediately
   * process the payment instead of setting it as default.
   */
  isRecycling?: boolean;
  /**
   * Callback fired when the user tries to close the modal by either
   * clicking on the close button, on the backdrop or by pressing the escape key
   */
  onClose: () => void;
  /** Callback fire when the user clicks the **Add Other Payment Methods** button. */
  onAddPaymentMethod?: (open: boolean) => void;
  /** Callback fired when a payment method was selected. */
  onSelectedPaymentMethod?: () => void;
  /** Default payment methods, if we fetch them outside the component */
  defaultPaymentMethods?: { payment_methods: PaymentMethods[] };
}) => {
  const [transitionInProgress, setTransitionInProgress] =
    useState<boolean>(false);
  const [swiper, setSwiper] = useState<Swiper | null>(null);
  const swiperSlideTo = (index: number) => swiper?.slideTo(index);

  let paymentMethods = defaultPaymentMethods;
  if (defaultPaymentMethods === undefined) {
    const { data } = useGetPaymentMethod();
    paymentMethods = data;
  }

  const { isError, refetch: refetchPaymentMethods } = useGetPaymentMethod();

  const defaultPaymentQuery = SetDefaultMethodPaymentTest(
    refetchPaymentMethods,
    swiperSlideTo
  );

  const processPaypalQuery = useProcessPayPal();
  const processCardQuery = useProcessCard();

  const {
    session: { account, isAuthenticated },
    useAccount,
  } = useSession();

  const { refetch: refetchAccount } = useAccount(isAuthenticated);

  const context = useContext(AppConstants);

  const processPaymentMethod = (
    slideData: Pick<PaymentMethods, 'id' | 'payment_type'>
  ) => {
    if (!account?.account.subscription_info?.subscription_plan_unique_name)
      return;
    const plan = {
      unique_key:
        account?.account.subscription_info?.subscription_plan_unique_name,
    };

    if (slideData.payment_type === 'paypal') {
      processPaypalQuery.mutate(
        {
          paymentAgreementId: slideData.id,
          plan,
        },
        {
          onSuccess: () => {
            refetchAccount();
            refetchPaymentMethods();
          },
        }
      );
    } else {
      processCardQuery.mutate(
        { creditCardId: slideData.id, plan },
        {
          onSuccess: () => {
            refetchAccount();
            refetchPaymentMethods();
          },
        }
      );
    }

    onClose();
  };

  return (
    <Modal
      title="Select from your payment methods"
      open={open}
      onClose={onClose}
      hasCloseIcon
      sx={{
        '& #modal-title': {
          fontSize: { xs: '16px', sm: '24px' },
          padding: { xs: '16px 24px 0px 24px', sm: '16px 24px' },
        },
      }}
    >
      {paymentMethods?.payment_methods?.length === 0 ||
      paymentMethods === undefined ||
      isError ? (
        <Text align="center" marginBottom={2}>
          No Payments associated to this account
        </Text>
      ) : (
        <>
          <FlexDivCenter>
            <Text
              variant="subtitle2"
              fontSize={{ xs: '14px', sm: '16px' }}
              gutterBottom
              className="text-support"
            >
              The payment method you select here will become your default
              payment for subscription payments and purchases
            </Text>
          </FlexDivCenter>
          <CarouselWrapper>
            <Carousel
              config={{
                slidesPerView: 1,
                spaceBetween: 150,
                speed: 500,
                loop: true,
                pagination: { clickable: true },
                navigation: {
                  nextEl: '.swiper-button-next',
                  prevEl: '.swiper-button-prev',
                },
                centeredSlides: true,
                modules: [Pagination, Navigation],
                className: 'payments-swiper',
                onSwiper: setSwiper,
                onSlideChangeTransitionStart: () =>
                  setTransitionInProgress(true),
                onSlideChangeTransitionEnd: () => {
                  setTransitionInProgress(false);
                },
              }}
              elements={carouselElements(paymentMethods, context)}
            />
          </CarouselWrapper>
        </>
      )}

      <FlexDivContainer
        sx={{ flexDirection: inReactivation ? 'column' : 'row' }}
      >
        <Button
          onClick={() => {
            if (!swiper) return;
            const activeSlide = swiper.slides[swiper.activeIndex];
            if (activeSlide instanceof HTMLElement) {
              const slideData = getSlideData(
                activeSlide.firstChild as HTMLElement
              );

              // TODO: should extract this logic to the parent component because here we are mixing concerns - RS 2023-06-01
              if (isRecycling) return processPaymentMethod(slideData);

              defaultPaymentQuery.mutate(Object.assign({}, slideData), {
                onSuccess: () => {
                  onSelectedPaymentMethod?.();
                },
              });
            }
          }}
          variant="contained"
          color="primary"
          disabled={transitionInProgress}
          sx={{
            fontSize: { xs: '14px', sm: '16px' },
          }}
        >
          {cta}
        </Button>

        <Button
          variant="contained"
          color="secondary"
          onClick={() => onAddPaymentMethod?.(true)}
          sx={{
            fontSize: { xs: '14px', sm: '16px' },
          }}
        >
          Add Other Payment Methods
        </Button>
      </FlexDivContainer>
    </Modal>
  );
};
