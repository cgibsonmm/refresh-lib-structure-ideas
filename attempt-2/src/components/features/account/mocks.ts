import { rest } from 'msw';
import { mockResolvers } from '@/utils/TestUtils';
import { fetchAccountMock } from './account.mock';
import {
  RESET_PASSWORD_ENDPOINT,
  REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT,
} from './constants';
import { FETCH_ACCOUNT_ENDPOINT } from './useAccount';
import { UPDATE_SETTING_ENDPOINT } from './useUserSetting';
import type { RestRequest, RestContext, ResponseResolver } from 'msw';

const requestResetPasswordEmailResponses = {
  success: {
    account: {
      message:
        'Great. We will send you instructions if an account with that email exists.',
    },
    meta: { status: 200 },
  },
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const fetchAccountResponses = {
  success: fetchAccountMock,
  badRequest: {
    account: { errors: ['Missing Parameters'] },
    meta: { status: 400 },
  },
};

const updateSettingResponses = {
  success: {
    message: 'Ok',
    meta: { status: 200 },
  },
  badRequest: {
    message: 'Bad Request',
    meta: { status: 400 },
  },
};

const resetPasswordResponses = {
  success: {},
  badRequest: {
    account: {
      messages: [
        'The reset token has expired. Please try to reset your password again.',
      ],
    },
    meta: { status: 400 },
  },
};

const sendCancellationRequestResponses = {
  success: {},
  badRequest: {
    account: {
      messages: ['Could not cancel account'],
    },
  },
};

const requestResetPasswordEmailResolvers = mockResolvers(
  requestResetPasswordEmailResponses,
  {
    default: async (req, res, ctx) => {
      const body = await req.arrayBuffer();
      const parsedBody = new URLSearchParams(body.toString());

      const isInvalidContentType =
        req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
      const isMissingBody = !parsedBody;
      // Because of how we're parsing the request body here, undefined email values were being converted to the string "undefined," and failing to trigger the error response. Adding an extra check for the literal string "undefined" solves the issue, but this may be an indication that we should be parsing 'x-www-form-urlencoded' requests differently in our mocks. - c.j.
      const isMissingEmail =
        parsedBody.get('account[email]') === (undefined || 'undefined');
      const isInvalidRequest =
        isInvalidContentType || isMissingBody || isMissingEmail;

      if (isInvalidRequest)
        return res(
          ctx.status(400),
          ctx.json(requestResetPasswordEmailResponses.badRequest)
        );
      else
        return res(
          ctx.status(200),
          ctx.json(requestResetPasswordEmailResponses.success)
        );
    },
  }
);

const resetPasswordResolvers = mockResolvers(resetPasswordResponses, {
  default: async (req, res, ctx) => {
    const validPassword = 'VALID_TEST_PASSWORD';
    const validToken = 'VALID_TEST_TOKEN';
    const body = await req.arrayBuffer();
    const parsedBody = new URLSearchParams(body.toString());
    const isInvalidContentType =
      req.headers.get('Content-Type') !== 'application/x-www-form-urlencoded';
    const hasInvalidParams =
      parsedBody.get('account[password]') !== validPassword ||
      parsedBody.get('account[password_confirmation]') !== validPassword ||
      parsedBody.get('password_reset_token') !== validToken;
    const isInvalidRequest = isInvalidContentType || hasInvalidParams;
    if (isInvalidRequest)
      return res(ctx.status(400), ctx.json(resetPasswordResponses.badRequest));
    else return res(ctx.status(200), ctx.json(resetPasswordResponses.success));
  },
});

const sendCancellationRequestResolvers = mockResolvers(
  sendCancellationRequestResponses,
  {
    default: async (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json(sendCancellationRequestResponses.success)
      );
    },
  }
);

const fetchAccountResolvers = mockResolvers(fetchAccountResponses);

type MockResolver = ResponseResolver<RestRequest<URLSearchParams>, RestContext>;

function requestResetPasswordEmailHandler(resolver: MockResolver) {
  return rest.post(REQUEST_RESET_PASSWORD_EMAIL_ENDPOINT, resolver);
}

function resetPasswordHandler(resolver: MockResolver) {
  return rest.post(RESET_PASSWORD_ENDPOINT, resolver);
}

function fetchAccountHandler(resolver: MockResolver) {
  const fetchAccountUrl = FETCH_ACCOUNT_ENDPOINT;
  return rest.get(fetchAccountUrl, resolver);
}

const updateSettingResolvers = mockResolvers(updateSettingResponses);

function updateSettingHandler(resolver: MockResolver) {
  const updateSettingUrl = UPDATE_SETTING_ENDPOINT;
  return rest.patch(updateSettingUrl, resolver);
}

function sendCancellationRequestHandler(resolver: MockResolver) {
  const sendCancellationRequestUrl = '/api/v5/account/cancel';
  return rest.post(sendCancellationRequestUrl, resolver);
}

export const handlers = {
  requestResetPasswordEmail: requestResetPasswordEmailHandler,
  resetPassword: resetPasswordHandler,
  fetchAccount: fetchAccountHandler,
  updateSetting: updateSettingHandler,
  sendCancellationRequest: sendCancellationRequestHandler,
};

export const resolvers = {
  requestResetPasswordEmail: requestResetPasswordEmailResolvers,
  resetPassword: resetPasswordResolvers,
  fetchAccount: fetchAccountResolvers,
  updateSetting: updateSettingResolvers,
  sendCancellationRequest: sendCancellationRequestResolvers,
};

export const responses = {
  requestResetPasswordEmail: requestResetPasswordEmailResponses,
  resetPassword: resetPasswordResponses,
  fetchAccount: fetchAccountResponses,
  updateSetting: updateSettingResponses,
  sendCancellationRequest: sendCancellationRequestResponses,
};
