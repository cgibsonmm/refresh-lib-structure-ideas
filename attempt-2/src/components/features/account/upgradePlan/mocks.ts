import { Plan, CurrentPlan } from './UpgradePlan';
import { costPerReport, savings } from './utils';

export const basePlanMock: Plan = {
  key: '26_89_1_month_nofree_afill_freshness_limit_100_freepdf_upsell',
  title: 'Starter',
  description: 'Starting small and simple',
  amount: 26.89,
  limit: 100,
  renewalPeriod: 1,
};

export const plansMock: Plan[] = [
  {
    key: '36_58_1_month_nofree_afill_freshness_limit_200_freepdf_upsell',
    title: 'Standard',
    description: 'Double your search capacity',
    amount: 36.58,
    limit: 200,
    renewalPeriod: 1,
    costPerReport: costPerReport(36.58, 1, 200),
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      36.58,
      1,
      200
    ),
  },
  {
    key: '63_99_1_month_nofree_afill_freshness_limit_400_freepdf_upsell',
    title: 'Premium',
    description: 'Your premium reports package',
    amount: 63.99,
    limit: 400,
    renewalPeriod: 1,
    costPerReport: costPerReport(63.99, 1, 400),
    mostPopular: true,
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      63.99,
      1,
      400
    ),
  },
  {
    key: '114_29_1_month_nofree_afill_freshness_limit_1000_freepdf_upsell',
    title: 'Professional',
    description: 'Take your business to the next level',
    amount: 114.29,
    limit: 1000,
    renewalPeriod: 1,
    costPerReport: costPerReport(114.29, 1, 1000),
    savings: savings(
      basePlanMock.amount,
      basePlanMock.renewalPeriod,
      basePlanMock.limit,
      114.29,
      1,
      1000
    ),
  },
];

export const currentPlanMock: CurrentPlan = {
  title: basePlanMock.title,
  limit: basePlanMock.limit,
  costPerReport: costPerReport(
    basePlanMock.amount,
    basePlanMock.renewalPeriod,
    basePlanMock.limit
  ),
};
