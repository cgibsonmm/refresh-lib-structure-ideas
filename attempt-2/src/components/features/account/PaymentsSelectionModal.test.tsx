import { render, screen } from '@testing-library/react';
import { AppConstants, defaultConstants } from '@/context/AppConstants';
import { getPaymentMethodsHandler } from '@/mocks/handlers';
import { server } from '@/mocks/server';
import { BrandThemeProvider, Modal } from '@/theme/index';
import { disableReactQueryErrorLogs, withQueryClient } from '@/utils/TestUtils';
import { PaymentsSelectionModal } from './PaymentsSelectionModal';

disableReactQueryErrorLogs();

const Component = () => {
  const onClose = jest.fn();

  return (
    <Modal open={true}>
      <PaymentsSelectionModal
        open
        onClose={onClose}
        defaultPaymentMethods={{ payment_methods: [] }}
      />
    </Modal>
  );
};

test('The PaymentSelection screen renders on the document', () => {
  server.use(getPaymentMethodsHandler);

  const { asFragment } = render(withProviders(Component));

  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText('Make Default')).toBeInTheDocument();
});

function withProviders(Component: () => JSX.Element) {
  return (
    <AppConstants.Provider value={defaultConstants}>
      <BrandThemeProvider>{withQueryClient({ Component })}</BrandThemeProvider>
    </AppConstants.Provider>
  );
}
