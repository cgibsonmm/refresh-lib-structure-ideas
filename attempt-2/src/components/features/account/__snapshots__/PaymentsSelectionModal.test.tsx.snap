// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PaymentSelection screen renders on the document 1`] = `<DocumentFragment />`;
